### ISSUE TITLE
Cannot read property 'domains' of undefined

### ISSUE THREAD

**Issue by flosoft** [nwesterhausen/domain-monitor#22]
**Describe the bug**
When adding a domain with whois caching, the page just keeps saying: `No WHOIS cache for domain.tld`
Running Docker Image: `nwesterhausen/domain-monitor`

```
domain-monitor    | (node:1) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'domains' of undefined
domain-monitor    |     at sendCachedWHOISData (/app/lib/sockets.js:25:14)
domain-monitor    |     at /app/lib/sockets.js:120:9
domain-monitor    | (node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
```

**To Reproduce**
- Run Docker image `nwesterhausen/domain-monitor`
- Add a 2 domains.
- For each add, the error above will be thrown and neither domains will update their whois data / status.

**Expected behavior**
The domain status should update.

**Screenshots**
<img width="1466" alt="Screenshot 2021-07-25 at 10 26 34" src="https://user-images.githubusercontent.com/602633/126892766-9d7b5ddd-11e7-4dd5-9d90-37a0fed1fdd4.png">

**Desktop (please complete the following information):**
- Docker on Ubuntu 20.04 LTS

**Comment by nwesterhausen** [comment]
I believe this is fixed, if you can verify. After making the changes I pulled in I no longer get the error you listed.

This update should be live on the docker hub image.

**Comment by flosoft** [comment]
Thanks - the error no longer seems to be popping up, but the domains don't seem to be updating from the second record onwards.
Steps:
- Add florianjensen.com
- Remove 2 demo domains
- Add jensen.com

The first domain will get the whois info stored and update status, but the second one does not.


Only other error I can find in the log (although I'm not sure it's related) is:
```
domain-monitor    | (node:1) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'map' of undefined
domain-monitor    |     at sendCachedWHOISData (/app/lib/sockets.js:25:22)
domain-monitor    |     at /app/lib/sockets.js:147:9
domain-monitor    | (node:1) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
```

I'm running it on: https://domain-monitor.jensen.ovh

**Comment by nwesterhausen** [comment]
It's probably another improperly coded promise chain on my part. I will duplicate on my end and see what I discover

**Comment by nwesterhausen** [comment]
Found the issue, when the whois data for jensen.com was being stored, it was being stored as it's listed in the whois result. In this case, it was all caps `JENSON.COM` but when I store references to and build the webpage for the domains, I lowercase them (or they can be lowercased). Anyway, after manually changing the `domain_name` value in `jensen.com.yaml` it loaded and so did the rest of the page. 

I modified the code to store `domain_name` in the cached whois data as lowercase, which should match what the frontend webpage expects to receive. After these changes, I was able to run the code fresh and follow your steps and have things successful.

**Comment by nwesterhausen** [comment]
docker images have been updated

**Comment by flosoft** [comment]
Awesome - thanks for the fixes! Really appreciate it.

Just added a bunch of domains - now waiting for the app to pull whois.

https://domain-monitor.jensen.ovh/#

**Comment by nwesterhausen** [comment]
I think it will catch them 24 hours later if you added the domains > 1min after starting the service. I don't think I have it set up to check the domains immediately after adding them. Once the domains are listed in domains.yaml, it will catch them on it's 1min check after startup.

**Comment by flosoft** [comment]
Alright - making progress :)
Sadly it seems that after 24 hours some of the domain names are just not loading anymore at all.

<img width="1230" alt="Screenshot 2021-08-01 at 17 27 15" src="https://user-images.githubusercontent.com/602633/127776490-a2f37fc2-7236-4bbe-b9b6-8e28c0ca37d9.png">

**Comment by nwesterhausen** [comment]
Does the web console list any errors?

**Comment by flosoft** [comment]
Sadly not. No errors in there :/

**Comment by nwesterhausen** [comment]
I got it to reproduce, so I will figure this out

![image](https://user-images.githubusercontent.com/2317381/127776855-848a765b-a49f-41c7-a0d2-2d30fbf4c473.png)

**Comment by nwesterhausen** [comment]
Fixed in 0.3.1 

There were some issues with reading files that somehow didn't surface before. I have tested myself and it seems fixed so let me know if this fixes the issues you're seeing.

**Comment by flosoft** [comment]
Still seems to be stuck. I think it could be to do with `jensen.eu` as the domain is the only one missing (probably `undefined`) in this log:
```
domain-monitor    | 2021-08-01T17:53:35.689521919Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | connection
domain-monitor    | 2021-08-01T17:53:35.694088947Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | sent config
domain-monitor    | 2021-08-01T17:53:35.696139777Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | sent domains
domain-monitor    | 2021-08-01T17:53:35.703677957Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for florianjensen.com
domain-monitor    | 2021-08-01T17:53:35.709208333Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for flosoft.biz
domain-monitor    | 2021-08-01T17:53:35.713062471Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for jensen.com
domain-monitor    | 2021-08-01T17:53:35.715339107Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for undefined
domain-monitor    | 2021-08-01T17:53:35.717463971Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for jensen.io
domain-monitor    | 2021-08-01T17:53:35.719582064Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for jensen.me
domain-monitor    | 2021-08-01T17:53:35.722852584Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for jensen.net
domain-monitor    | 2021-08-01T17:53:35.725899147Z ::ffff:172.28.0.5 - WS client sAVlAvOF-NnkSlj_AAAN | Sent WHOIS for jensen.org
```

I've attached the config and whois files as well as my docker-compose.yml here.
[dm-backup.tar.gz](https://github.com/nwesterhausen/domain-monitor/files/6913134/dm-backup.tar.gz)

**Comment by flosoft** [comment]
Confirmed that removing `jensen.eu` does the trick.

**Comment by nwesterhausen** [comment]
I think I see the problem already. I'm doing to try running the fix on your provided backup (thanks!).

**Comment by nwesterhausen** [comment]
The parser does not like how the whois comes back from the eu domain. I'll probably write a second parser and direct the results there for eu (and any other found in the future which don't follow the convention of `^key: value$`

**Comment by nwesterhausen** [comment]
Should be fixed, published to :latest and :0.3.2.

I recommend removing `jensen.eu.yaml` from your whois-data so that we know it will re-create it. The error that was occurring was happening because the parser was unable to find a domain_name to put into the yaml and that broke things. If it can't find a domain_name in the WHOIS data now, it will just return empty and the server already knows to handle that (it will put an error in the log for this).

I didn't write a new parser, just a per-processor that dissects the WHOIS data if it detects indented values.
